Custom pid errors

cowsed opened the pull request:
Add custom error functions to pid class. 
Error calculation functions are of the form  `double calculate_error(double sensor_val, double target)`. 
Useful in situations where the error is not strictly `target - sensor_val` (angles).
If no custom function is specified, `target - sensor_val` is still used.

superrm11 committed:
Restructured odometry asynchronous model

superrm11 committed:
Added Odometry3Wheel

superrm11 committed:
Fixed core.h, added testing code

superrm11 committed:
Added tuning helper, fixed math error

cowsed committed:
Created a new constructor for PID that allows specifying a custom err or function with a function pointer.
If no function pointer is provided, the normal `target - sensor_val` calculation is used

cowsed committed:
Added documentation for custom errors in pid
New Constructors for pid to take a custom error
New Constructors for motion profile to take a custom error

cowsed committed:
Edited get_error() to actually use the custom error function
Somehow forgot that in the last commit

dec1701 commented on core/include/utils/motion_controller.h:
```diff
@@ -43,6 +43,17 @@ class MotionController : public Feedback
+    /**
+     * @brief Construct a new Motion Controller object
+     * 
+     * @param max_v Maximum velocity the movement is capable of
```
can we make it clear that these parameters are passed through the m_profile_cfg_t struct?

cowsed replied:
yes

cowsed committed:
Clarified motion profile constructor.

cowsed committed:
added one space to make it all align

superrm11 commented:
If this is an issue with strictly angles, it might be better to simplify with an enum option, like:
`PID pid(cfg, Type::LINEAR); //or Type::ANGULAR`
and let the PID or Motion Profile class handle it. Might be less confusing than a function pointer, and invite less error.  Thoughts?

cowsed commented:
Yeah - and if a need arises for a different custom strategy thats just another enum value.

Rabinvic committed:
Replaced pointer with enum

Replaced the function pointer in the PID class with an enum for calculation method

cowsed committed:
Removed comment referring to function pointer

cowsed committed:
Removed unused PID constructor from PID.h

superrm11 committed:
added angle_wrap

superrm11 committed:
Fixed issue with 3pod odometry

superrm11 committed:
Merge pull request #7 from RIT-VEX-U/motion_profile_dev

Motion profile dev

superrm11 committed:
Merge remote-tracking branch 'origin/main' into odometry

superrm11 committed:
Merge pull request #10 from RIT-VEX-U/odometry

Odometry

superrm11 committed:
Merge branch 'main' into custom-pid-errors
